elephant/spike_train_correlation.py

```diff
@@ -0,0 +1,131 @@
+# -*- coding: utf-8 -*-
+"""
+Spike train correlation
+
+This modules provides functions to calculate correlations between spike trains.
+
+:copyright: Copyright 2014 by the Elephant team, see AUTHORS.txt.
+:license: Modified BSD, see LICENSE.txt for details.
+"""
+from __future__ import division
+import numpy as np
+
+
+def corrcoef(binned_sts, binary=False):
+    '''
+    Calculate the NxN matrix of pairwise Pearson's correlation coefficients
+    between all combinations of N binned spike trains.
+
+    For each pair of spike trains $(i,j)$, the correlation coefficient $C[i,j]$
+    is given by the correlation coefficient between the vectors obtained by
+    binning $i$ and $j$ at the desired bin size. Let $b_i$ and $b_j$ denote the
+    binary vectors and $m_i$ and  $m_j$ their respective averages. Then
+
+    .math $$ C[i,j] = <b_i-m_i, b_j-m_j> /
+                      \sqrt{<b_i-m_i, b_i-m_i>*<b_j-m_j,b_j-m_j>} $$
+
+    where <..,.> is the scalar product of two vectors.
+
+    For an input of n spike trains, a n x n matrix is returned.
+    Each entry in the matrix is a real number ranging between -1 (perfectly
+    anti-correlated spike trains) and +1 (perfectly correlated spike trains).
+
+    If binary is True, the binned spike trains are clipped before computing the
+    correlation coefficients, so that the binned vectors b_i, b_j are binary.
+
+    Parameters
+    ----------
+    binned_sts : elephant.conversion.BinnedSpikeTrain
+        A binned spike train class containing the spike trains to be evaluated.
+    binary : bool, optional
+        If True, two spikes of a particular spike train falling in the same
+        bin are counted as 1, resulting in binary binned vectors b_i. If False,
+        the binned vectors $b_i$ contain the actually spike counts per bin.
+        Default: False
+
+    Returns
+    -------
+    C : ndarrray
+        The square matrix of correlation coefficients. The element
+        $C[i,j]=C[j,i]$ is the Pearson's correlation coefficient between
+        binned_sts[i] and binned_sts[j]. If binned_sts contains only one
+        SpikeTrain, C=1.0.
+
+    Notes
+    -----
+    * The spike trains in the binned structure are assumed to all cover the
+      complete time span of binned_sts [t_start,t_stop).
+    '''
+    num_neurons = binned_sts.matrix_rows
+
+    # Pre-allocate correlation matrix
+    C = np.zeros((num_neurons, num_neurons))
+
+    # Retrieve unclipped matrix
+    spmat = binned_sts.to_sparse_array()
+
+    # For each row, extract the nonzero column indices and the corresponding
+    # data in the matrix (for performance reasons)
+    bin_idx_unique = []
+    bin_counts_unique = []
+    if binary:
```

Comment:
I was wondering if it would be more cleaner in terms of code complexity to split this function internally into 2 private functions which are called depending on the parameter `binary`.